internal/database/functions.go

```diff
@@ -0,0 +1,74 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package database
+
+import (
+	"database/sql/driver"
+	"errors"
+	"fmt"
+
+	"github.com/Masterminds/semver/v3"
+	"modernc.org/sqlite"
+)
+
+func init() {
+	sqlite.MustRegisterScalarFunction("semver_compare_constraint", 2, semverCompareConstraint)
+	sqlite.MustRegisterScalarFunction("semver_compare_op", 3, semverCompareOperation)
+}
+
+// semverCompare checks if a version satisfies a given semver constraint.
+// It takes two string arguments: the version and the constraint.
+// It returns a boolean indicating whether the version satisfies the constraint.
+func semverCompareConstraint(ctx *sqlite.FunctionContext, args []driver.Value) (driver.Value, error) {
+	version, ok := args[0].(string)
+	if !ok {
+		return nil, errors.New("version argument must be a string")
+	}
+	versionSemver, err := semver.NewVersion(version)
+	if err != nil {
+		return nil, err
+	}
+
+	constraint, ok := args[1].(string)
+	if !ok {
+		return nil, errors.New("constraint argument must be a string")
+	}
+
+	// If there is no constraint provided, consider it a match
+	if constraint == "" {
+		return true, nil
+	}
+
+	constraintSemver, err := semver.NewConstraint(constraint)
+	if err != nil {
+		return nil, fmt.Errorf("invalid semver constraint: %w", err)
+	}
+
+	return constraintSemver.Check(versionSemver), nil
+}
+
+// semverCompareOperation compares two semantic versions based on a given operation.
+// It takes three string arguments: the first version, the operation (e.g., '>=', '<=', etc.), and the second version.
+// It returns a boolean indicating whether the comparison holds true.
+func semverCompareOperation(ctx *sqlite.FunctionContext, args []driver.Value) (driver.Value, error) {
+	firstVersion, ok := args[0].(string)
+	if !ok {
+		return nil, fmt.Errorf("firstVersion argument must be a string")
+	}
+
+	operation, ok := args[1].(string)
+	if !ok {
+		return nil, fmt.Errorf("operation argument must be a string")
+	}
+
+	secondVersion, ok := args[2].(string)
+	if !ok {
+		return nil, fmt.Errorf("secondVersion argument must be a string")
+	}
+
+	newArgs := []driver.Value{firstVersion, fmt.Sprintf("%s%s", operation, secondVersion)}
+
+	return semverCompareConstraint(ctx, newArgs)
```

Comment:
Compiling constraints is relatively expensive, as we saw in some other profiling. 

Maybe it is better to have two different functions, `semver_compare_le` and `semver_compare_ge` for less than and greater than respectively, that would only require to compile the two versions, and use their methods.

Reply:
I'll add those two new functions :+1: 
Even it is better since using constraints there are some caveats when using pre-release versions, since by default they are excluded:
https://github.com/Masterminds/semver?tab=readme-ov-file#checking-version-constraints

> When constraint checking is used for checks or validation it will follow a different set of rules that are common for ranges with tools like npm/js and Rust/Cargo. This includes considering pre-releases to be invalid if the ranges does not include one. If you want to have it include pre-releases a simple solution is to include -0 in your range.